File: ui-preview/src/androidTest/java/l/files/thumbnail/PdfThumbnailerTest.java
package l.files.thumbnail;

import android.content.res.AssetManager;

import org.junit.Test;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;

import l.files.fs.Path;
import l.files.testing.fs.Paths;
import l.files.ui.base.graphics.Rect;
import l.files.ui.base.graphics.ScaledBitmap;

import static android.support.test.InstrumentationRegistry.getContext;
import static java.io.File.createTempFile;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

public final class PdfThumbnailerTest {

    @Test
    public void create_thumbnail_from_pdf() throws Exception {
        Path path = createTestPdf();
        try {
            Rect max = Rect.of(10, 100);
            ScaledBitmap result = newThumbnailer().create(path, max, getContext());
            assertNotNull(result);
            assertEquals(max.width(), result.bitmap().getWidth());
            assertTrue(result.originalSize().width() > max.width());
            assertTrue(result.originalSize().height() > max.height());
        } finally {
            Paths.deleteIfExists(path);
        }
    }

    private Path createTestPdf() throws IOException {
        File file = createTempFile("PdfThumbnailerTest", null);
        try {
            Path path = Path.of(file);
            try (InputStream in = openTestPdf()) {
                Paths.copy(in, path);
            }
            return path;
        } catch (Throwable e) {
            assertTrue(file.delete() || !file.exists());
            throw e;
        }

    }

    private InputStream openTestPdf() throws IOException {
        return getAssets().open("PdfThumbnailerTest.pdf");
    }

    private AssetManager getAssets() {
        return getContext().getAssets();
    }

    private PdfThumbnailer newThumbnailer() {
        return new PdfThumbnailer();
    }

}

File: ui-preview/src/androidTest/java/l/files/thumbnail/PdfThumbnailerTest.kt
package l.files.thumbnail

import androidx.test.platform.app.InstrumentationRegistry.getInstrumentation
import l.files.ui.base.graphics.Rect
import org.junit.Assert.*
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TemporaryFolder
import java.io.IOException
import java.nio.charset.StandardCharsets.UTF_8
import java.nio.file.Files.*
import java.nio.file.Path
import java.nio.file.StandardCopyOption.REPLACE_EXISTING

class PdfThumbnailerTest {

  @Rule
  @JvmField
  val folder = TemporaryFolder()

  private val context get() = getInstrumentation().context
  private val assets get() = context.assets

  @Test
  fun create_mix_failures() {
    // https://issuetracker.google.com/issues/37016953

    val valid = createTestPdf()
    val invalid = createInvalidPdf()
    val max = Rect.of(10, 100)

    PdfThumbnailer.create(valid, max, context)

    try {
      PdfThumbnailer.create(invalid, max, context)
      fail()
    } catch (e: IOException) {
      // Pass
    }

    PdfThumbnailer.create(valid, max, context)
  }

  @Test
  fun create_thumbnail_from_pdf() {
    val path = createTestPdf()
    try {
      val max = Rect.of(10, 100)
      val result = PdfThumbnailer.create(path, max, context)!!
      assertEquals(
        max.width().toLong(),
        result.bitmap().width.toLong()
      )
      assertTrue(result.originalSize().width() > max.width())
      assertTrue(result.originalSize().height() > max.height())
    } finally {
      deleteIfExists(path)
    }
  }

  private fun createTestPdf(): Path {
    val path = folder.newFile("PdfThumbnailerTest").toPath()
    openTestPdf().use { copy(it, path, REPLACE_EXISTING) }
    return path
  }

  private fun createInvalidPdf(): Path {
    val path = folder.newFile("PdfThumbnailerTestInvalid").toPath()
    write(path, "hello".toByteArray(UTF_8))
    return path
  }

  private fun openTestPdf() = assets.open("PdfThumbnailerTest.pdf")

}
